src/pip/_internal/req/req_install.py

```diff
@@ -257,7 +257,7 @@ def from_line(
                 elif '=' in req and not any(op in req for op in operators):
                     add_msg = "= is not a valid operator. Did you mean == ?"
                 else:
-                    add_msg = traceback.format_exc()
+                    add_msg = ''
```

Comment:
As I mentioned in https://github.com/pypa/pip/issues/5147#issuecomment-378001502, it's probably not right to print the traceback or nothing here.

Could you look into extracting any useful info from the exception raised?

Reply:
Hi @pradyunsg thanks for the review!

Well, we have 
```python
pip._vendor.pyparsing.ParseException: Expected W:(abcd...) (at char 0), (line:1, col:1)
```
 at the end of the first stack trace and
```python
pip._vendor.packaging.requirements.InvalidRequirement: Invalid requirement, parse error at "'-'"
```
at the end of the second one.

And of course the error message
```
Invalid requirement: '-'
```
I think for a user facing error message the `Invalid requirement: '-'` should be explanation enough for a user to fix the problem. The extra info about the actual problem -- the parser error -- does not really add value don't you think?

Looking at the issue it seems that @gvanrossum and @pfmoore had a similar fix in mind.

Reply:
I agree with @pradyunsg. I think it would be better to display something about _why_ / what caused the requirement to be invalid as opposed to swallowing the details completely and just saying it's "invalid." For now this could be done simply by displaying the exception type and message for what caused it. For the case that was originally provided, this could look something like--

    Invalid requirement: '-'
    Details: ParseException: Expected W:(abcd...) (at char 0), (line:1, col:1)

Over time, these inner causes could be made more user-friendly as they come up. For example, in this case it might be something like, "First character must be one of ...." (or whatever the condition is).

Reply:
Very well, I've updated the PR and it prints out now:
```bash
$ pip install "toto 42"
Invalid requirement: 'toto 42'
Expected stringEnd (at char 5), (line:1, col:6)
```
and
```bash
$ pip install -
Invalid requirement: '-'
Expected W:(abcd...) (at char 0), (line:1, col:1)
```
I'm not sure if value of the extra info justifies the fragility of the implementation to get it: Python2 doesn't support the `__context__` attribute yet and peeling out the message from the middle of a chained stack trace then seems not easy. So on Python2 it will simply print the `Invalid requirement` message without the additional hint.
```bash
$ pip install -
Invalid requirement: '-'
```

Reply:
The fact that you need to reach into the `__context__` makes it seem like something else is wrong. The info should already be contained in the message for the `InvalidRequirement` exception that you are handling. If it's not, then it seems like that message should be adjusted. This would also address the issue for Python 2 because then you'll no longer need to access `__context__` which isn't available. It will be available at the immediate exception being handled.

Reply:
@cjerdonek Ok, that'll make things a lot easier, I wasn't sure if I'm allowed to modify stuff in `src/pip/_vendor/`

Reply:
I just investigated, too. It looks like an issue would need to be filed with the [packaging](https://github.com/pypa/packaging) project to make that error message more informative. It's `packaging` that would need to provide a more detailed explanation since it's currently not providing that info:
https://github.com/pypa/pip/blob/0007825733aa862e76a66e5d1cfee121c29192ff/src/pip/_vendor/packaging/requirements.py#L95

So I guess @pradyunsg's option (2) is probably best here (combined with filing an issue with packaging).

Reply:
PS - just to be clear, you're probably not allowed to modify what's in the `_vendor` directory (which is why it would need to be improved in the project itself). Stuff in the vendor directory is most likely copied as is from wherever it came from.

Reply:
Ok, how about the current compromise:
```bash
$ pip install -
Invalid requirement: '-'
parse error at "'-'"
```
and
```bash
$ pip install 'toto 42'
Invalid requirement: 'toto 42'
parse error at "'42'"
```
The patch is very digestible and runs on all current Python's.